### PR TITLE
Implement Validators.ValidateIssuerSecurityKey

### PR DESCRIPTION
**Pull request by ChenYan98**
#329 on 4.x

### CODE REVIEW

File: src/common/ErrorMessages.cs [hunk 1]
```diff
@@ -74,6 +74,7 @@ public static class ErrorMessages
         public const string IDX10229 = "IDX10229: TokenValidationParameters.TokenReplayCache was unable to add the securityToken: '{0}'.";
         public const string IDX10230 = "IDX10230: Lifetime validation failed. Delegate returned false, securitytoken: '{0}'.";
         public const string IDX10231 = "IDX10231: Audience validation failed. Delegate returned false, securitytoken: '{0}'.";
+        public const string IDX10232 = "IDX10232: validationParameters.CertificateValidator is null; User need to set CertificateValidator";
```
Comment (brentschmaltz) [inline src/common/ErrorMessages.cs]: please change - "...; User need to set CertificateValidator" to "...; Validation of a X509SecurityKey requires that you set the CertificateValidator"

File: src/common/ErrorMessages.cs [hunk 2]
```diff
@@ -169,6 +170,7 @@ public static class ErrorMessages
         public const string IDX11006 = "IDX11006: This method is not supported to read a 'saml2token' use the method: ReadToken(string securityToken, TokenValidationParameters validationParameters).";
         public const string IDX11007 = "IDX11007: This method is not supported to read a 'samltoken' use the method: ReadToken(string securityToken, TokenValidationParameters validationParameters).";
         public const string IDX11008 = "IDX11008: This method is not supported to validate a 'jwt' use the method: ValidateToken(String, TokenValidationParameters, out SecurityToken).";
+        public const string IDX11009 = "IDX11009: This method only support X509SecurityKey";
```
Comment (brentschmaltz) [inline src/common/ErrorMessages.cs]: typo - support should be supports.

File: src/System.IdentityModel.Tokens.Jwt/Validators.cs [hunk 3]
```diff
@@ -169,8 +169,14 @@ public static void ValidateIssuerSecurityKey(SecurityKey securityKey, SecurityTo
             X509SecurityKey x509SecurityKey = securityKey as X509SecurityKey;
             if (x509SecurityKey != null)
             {
-                //validationParameters.CertificateValidator.Validate(x509SecurityKey.Certificate);
+                if (validationParameters.CertificateValidator == null)
+                    throw new NullReferenceException(ErrorMessages.IDX10232);
```
Comment (brentschmaltz) [inline src/System.IdentityModel.Tokens.Jwt/Validators.cs]: Throw a SecurityTokenValidationException here and below. NullReferenceException is normally thrown by the runtime.

File: src/System.IdentityModel.Tokens.Jwt/Validators.cs [hunk 4]
```diff
@@ -169,8 +169,14 @@ public static void ValidateIssuerSecurityKey(SecurityKey securityKey, SecurityTo
...
+
+                validationParameters.CertificateValidator.Validate(x509SecurityKey.Certificate);
+                return;
             }
+
+            throw new ArgumentOutOfRangeException("securityKey", ErrorMessages.IDX11009);
```
Comment (brentschmaltz) [inline src/System.IdentityModel.Tokens.Jwt/Validators.cs]: You need to add tests.

Comment (polita) [inline src/System.IdentityModel.Tokens.Jwt/Validators.cs]: I recommend changing this code flow around to make it easier to read and more maintainable.

`X509SecurityKey x509SecurityKey = securityKey as X509SecurityKey;`
`if (x509SecurityKey == null)`
`    throw new ArgumentOutOfRangeException("securityKey", ErrorMessages.IDX11009);`
`else `
`{`
`    ...`
`}`

File: src/common/ErrorMessages.cs (hunk 1 above)
Comment (polita) [inline src/common/ErrorMessages.cs]: Please also use a period instead of a semicolon to be consistent with the rest of the messages.

Comment (ChenYan98) [inline src/common/ErrorMessages.cs]: Modified

File: src/common/ErrorMessages.cs (hunk 2 above)
Comment (ChenYan98) [inline src/common/ErrorMessages.cs]: Modified

File: src/System.IdentityModel.Tokens.Jwt/Validators.cs (hunk 4 above)
Comment (ChenYan98) [inline src/System.IdentityModel.Tokens.Jwt/Validators.cs]: Modified

File: src/common/ErrorMessages.cs (hunk 1 above)
Comment (polita) [inline src/common/ErrorMessages.cs]: Have you updated your commit? I don't see the text modified.

File: src/common/ErrorMessages.cs (hunk 2 above)
Comment (polita) [inline src/common/ErrorMessages.cs]: I don't see this one modified either.

File: src/common/ErrorMessages.cs (hunk 1 above)
Comment (ChenYan98) [inline src/common/ErrorMessages.cs]: I was adding tests and updated my commit few minutes ago. Please check the latest commit. Thanks.

File: src/common/ErrorMessages.cs (hunk 2 above)
Comment (ChenYan98) [inline src/common/ErrorMessages.cs]: Same as above.

File: src/System.IdentityModel.Tokens.Jwt/Validators.cs (hunk 4 above)
Comment (ChenYan98) [inline src/System.IdentityModel.Tokens.Jwt/Validators.cs]: Tests added

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs [hunk 5]
```diff
@@ -70,9 +70,38 @@ public void CrossToken_ValidateToken()
...
+            };
+
+            // Test ValidateIssuerSigningKey = true, validationParameters.CertificateValidator == null
+            string jwtTokenTest = IdentityUtilities.CreateJwtToken(descriptor, jwtHandler);
+            ClaimsPrincipal jwtPrincipalTest = ValidateToken(jwtTokenTest, validationParameters, jwtHandler, ExpectedException.SecurityTokenValidationException("IDX10232:"));
+
             string nullIssuerJwtToken = IdentityUtilities.CreateJwtToken(IdentityUtilities.NullIssuerAsymmetricSecurityTokenDescriptor, jwtHandler);
             string jwtToken = IdentityUtilities.CreateJwtToken(IdentityUtilities.DefaultAsymmetricSecurityTokenDescriptor, jwtHandler);
 
+            // Test ValidateIssuerSigningKey = true, signingkey is not X509SecurityKey
```
Comment (brentschmaltz) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: We need a successful test.

File: src/System.IdentityModel.Tokens.Jwt/Validators.cs [hunk 6]
```diff
@@ -169,7 +169,14 @@ public static void ValidateIssuerSecurityKey(SecurityKey securityKey, SecurityTo
             X509SecurityKey x509SecurityKey = securityKey as X509SecurityKey;
             if (x509SecurityKey != null)
             {
-                //validationParameters.CertificateValidator.Validate(x509SecurityKey.Certificate);
+                if (validationParameters.CertificateValidator == null)
+                    throw new SecurityTokenValidationException(ErrorMessages.IDX10232);
+
+                validationParameters.CertificateValidator.Validate(x509SecurityKey.Certificate);
+            }
+            else
```
Comment (brentschmaltz) [inline src/System.IdentityModel.Tokens.Jwt/Validators.cs]: We need to check all key types. 
X509SecurityKey:
1. CertificateValidator 
2. IssuerSigningKeyValidator
Throw if both are null.

All other types.
use IssuerSigningKeyValidator, throw if null.

Comment (ChenYan98) [inline src/System.IdentityModel.Tokens.Jwt/Validators.cs]: Modified

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs (hunk 5 above)
Comment (ChenYan98) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: There is no validated x509SecurityKey in our project and I don't know how to create one. Except thisone, I added positive test cases for all the other path.

Comment (ChenYan98) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: There is no validated x509SecurityKey in our project. Do you have any idea to create a validated one? Except this, I added positive test cases for all the other paths.

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs [hunk 7]
```diff
@@ -70,9 +70,46 @@ public void CrossToken_ValidateToken()
...
+                AuthenticationType = IdentityUtilities.DefaultAuthenticationType,
+                IssuerSigningKey = x509SecurityKey,
+                ValidAudience = IdentityUtilities.DefaultAudience,
+                ValidIssuer = IdentityUtilities.DefaultIssuer,
+                ValidateIssuerSigningKey = true
+            };
+
+            // Test ValidateIssuerSigningKey = true, validationParameters.CertificateValidator == null, validationParameters.IssuerSigningKeyValidator == null
+            string jwtTokenTest = IdentityUtilities.CreateJwtToken(descriptor, jwtHandler);
+            ClaimsPrincipal jwtPrincipalTest = ValidateToken(jwtTokenTest, validationParameters, jwtHandler, ExpectedException.SecurityTokenValidationException("IDX10232:"));
```
Comment (polita) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: You don't ever use or check the value of jwtPrincipalTest. Do you need it?

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs [hunk 8]
```diff
@@ -70,9 +70,46 @@ public void CrossToken_ValidateToken()
...
+
+            // Test ValidateIssuerSigningKey = true, validationParameters.CertificateValidator == null, validationParameters.IssuerSigningKeyValidator != null
+            validationParameters.IssuerSigningKeyValidator = IssuerSigningKeyValidator;
+            jwtPrincipalTest = ValidateToken(jwtTokenTest, validationParameters, jwtHandler, ExpectedException.NoExceptionExpected);
+
             string nullIssuerJwtToken = IdentityUtilities.CreateJwtToken(IdentityUtilities.NullIssuerAsymmetricSecurityTokenDescriptor, jwtHandler);
             string jwtToken = IdentityUtilities.CreateJwtToken(IdentityUtilities.DefaultAsymmetricSecurityTokenDescriptor, jwtHandler);
 
+            // Test ValidateIssuerSigningKey = true, signingkey is not X509SecurityKey, alidationParameters.IssuerSigningKeyValidator == null
```
Comment (polita) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: Typo: validationParameters

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs [hunk 9]
```diff
@@ -70,9 +70,46 @@ public void CrossToken_ValidateToken()
...
+            validationParameters = IdentityUtilities.DefaultAsymmetricTokenValidationParameters;
+            validationParameters.ValidateIssuerSigningKey = true;
+            jwtPrincipalTest = ValidateToken(jwtToken, validationParameters, jwtHandler, ExpectedException.SecurityTokenValidationException("IDX10233:"));
+
+            // Test ValidateIssuerSigningKey = true, signingkey is not X509SecurityKey, alidationParameters.IssuerSigningKeyValidator != null
```
Comment (polita) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: Typo: validationParameters

Comment (polita) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: It's good to explicitly set each field for each test. That way, if someone else adds tests in between yours, they won't break your tests.

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs [hunk 10]
```diff
@@ -113,6 +150,10 @@ public void CrossToken_ValidateToken()
             JwtSecurityTokenHandler.InboundClaimFilter.Clear();
         }
 
+        private static void IssuerSigningKeyValidator(SecurityKey key)
```
Comment (polita) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: Do you ever validate that this gets called?

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs (hunk 9 above)
Comment (brentschmaltz) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: +1

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs (hunk 7 above)
Comment (brentschmaltz) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: I don't think this method is the right place for these tests for SecurityKey validation. These tests are for ensuring Saml and Jwt tokens return the same ClaimsPrincipal. 

Maybe here:
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev4x/tests/System.IdentityModel.Tokens.Jwt.Tests/ValidatorsTests.cs#L141

Comment (ChenYan98) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: @brentschmaltz Yes, I agree with you and I realized this yesterday. Please check it in my last commit.
@polita There is no need to create jwtPrincipalTest in the new method.

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs (hunk 8 above)
Comment (ChenYan98) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: Modified in new location.

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs (hunk 9 above)
Comment (ChenYan98) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: Same as above

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs (hunk 8 above)
Comment (ChenYan98) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: Modified in new location.

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs (hunk 9 above)
Comment (ChenYan98) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: Modified

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs (hunk 7 above)
Comment (ChenYan98) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: @brentschmaltz Yes, I agree with you and I realized this yesterday. In my last commit, I moved these tests to ValidatorsTests.Validators_SecurityKey()
@polita There is no need to create jwtPrincipalTest in the new method.

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs (hunk 9 above)
Comment (ChenYan98) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: Modified.

File: tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs (hunk 10 above)
Comment (ChenYan98) [inline tests/Microsoft.IdentityModel.Protocol.Extensions.Tests/CrossTokenTests.cs]: Set issuerSigningKeyValidatorCalled to validate this method get called.

**Comment by ChenYan98** [comment]
Leave a comment

**Pull request merged**